File: LeetCode-Java/Reorder List.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public void reorderList(ListNode head) {
        List<ListNode> record = new ArrayList<ListNode>();
        while(head != null)
        {
            record.add(head);
            head = head.next;
        }
        if(record.size() <= 1) return;
        int startIndex = 0, endIndex = record.size() - 1;
        int turn = 1;
        ListNode current = null, pre = null;
        while(startIndex <= endIndex)
        {
            int tempIndex = startIndex;
            if(turn == 0)
            {   
                tempIndex = endIndex;
                endIndex --;
            }
            else
                startIndex ++;
            current = record.get(tempIndex);
            if(pre != null) pre.next = current;
            pre = current;
            current.next = null;
            turn = 1 - turn;
        }
        return;
    }
}

File: Leetcode-cpp/reorder_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        if (head == nullptr) return;
        ListNode* slow = head;
        ListNode* fast = head;
        while (fast->next != nullptr && fast->next->next != nullptr) {
            fast = fast->next->next;
            slow = slow->next;
        }
        ListNode* right = reverseList(slow->next);
        slow->next = nullptr;
        ListNode* left = head;
        ListNode* fake_head = new ListNode(-1);
        ListNode* current = fake_head;
        while (left != nullptr) {
            current->next = left;
            current = current->next;
            left = left->next;;
            if (right != nullptr) {
                current->next = right;
                current = current->next;
                right = right->next;
            }
        }
        
    }
    ListNode* reverseList(ListNode* head) {
        ListNode* pre = nullptr;
        ListNode* current = head;
        while (current != nullptr) {
            ListNode* next = current->next;
            current->next = pre;
            pre = current;
            current = next;
        }
        return pre;
    }
};